### ISSUE TITLE
Flush could delete entities before inserting new ones

### ISSUE THREAD

**Issue by MacDada** [doctrine/doctrine2#5742]
I have a unique (not PK) column. I need to remove old entities first and then add new ones (replacements).

Yeah, I've seen this: https://github.com/doctrine/doctrine2/issues/5368

But I've also seen a "could be done" [solution](https://github.com/doctrine/doctrine2/issues/5109#issuecomment-162370533):

> A function "setShouldPerformDeletesFirst" as described in http://docs.oracle.com/cd/B31017_01/web.1013/b28218/uowadv.htm#i1137484 would be very useful in those cases. 

What I propose, is that this gets implemented. From the dev point of view, it would be cool, I one could do that like this:

`* @ORM\OneToMany(targetEntity="Whatever", mappedBy="Me", cascade={"persist"}, orphanRemoval=true, removeBeforeAdd=true)`

**Comment by ureimers** [comment]
I'm not quite sure if `removeBeforeAdd` should only be available on relations. You could easily think of a case where you replace entity A with B with both having the same uniquely constrained fields.

**Comment by IMM0rtalis** [comment]
+1 here

**Comment by johandouma** [comment]
+1. 
This issue seems related to #6776 - in case other people are searching.

**Comment by SmasherHell** [comment]
+1 That is quite a challenge to respect OCP when doctrine removes "merge". There is no way to update an object from a clone within a transaction. How to have a separation of concerns between logic and persistance when we must flush a state in the middle of logic

**Comment by hhamon** [comment]
I faced this issue as well on my client project. We had to remove the SQL unique constraint on the table to ensure the Doctrine SQL transaction can be complete. But ideally we would need both the SQL unique constraint (to always ensure consistency in the database) and in the SQL transaction when Doctrine flush statement is performed.

**Comment by mpdude** [comment]
@SmasherHell I am not sure if it would support your use case – probably it may be difficult to deal with two entities at the same time that share the same primary key value, and to have one replace the other in a single transaction.

The OP request is more about being able to meet other (non-primary key) constraints.

**Comment by mpdude** [comment]
#10809 adds an example that shows why we cannot – at least not, in general – change the current order of operations in the UoW.

**Comment by maddanio** [comment]
may I ask why the order is not simply preserved? that would likely be the expectation of the library user. doctrine reordering actions is susprising.

**Comment by mpdude** [comment]
No, that’s the promise given by the ORM: You tell it which entities you want to have tracked, and when you call `flush()`, it executes the (minimum amount of) queries it takes to persist changes.

That includes giving the ORM some freedom to order operations as it sees fit.

Additional unique constraints in the database as we‘re discussing here may interfere with that. They prevent the ORM from doing things in the planned order, through restrictions it cannot foresee and/or efficiently include in the commit order computation.

**Comment by maddanio** [comment]
can you give an example where not keeping the order is neccessary to fulfil that promise?

**Comment by greg0ire** [comment]
@maddanio you can read up on this at https://github.com/doctrine/orm/pull/10547

**Comment by maddanio** [comment]
hmm, but i don't quite see how this completely prevents keeping ordering. yes in some situations things need to be reordered, but if one really only does the required reorderings and keeps things otherwise stable I think most of the problems would go away, or am I overlookling something?

**Comment by mpdude** [comment]
> but if one really only does the required reorderings and keeps things otherwise stable I think most of the problems would go away

I think that is how it is implemented. If you have a reproducible case that indicates otherwise, I’d be happy to look into it as a potential bug

**Comment by maddanio** [comment]
Ah ok, well need to find out how to reduce this. But basically i am removing an entity and then adding another one with the same primary id. Adding a flush between makes it work. Hmm, but maybe we are using a too old doctrine version. Will check

**Comment by mpdude** [comment]
You’re even describing a special case where the unique constraint is for the primary key, not just for any other column.

You must be using application-provided IDs, otherwise this kind of ID reuse would not happen.

This is not going to work for at least one reason apart from the order of operations during flush: Since the identity map refers to entities (object instances) by means of the ID, there will be a collision already when you try to add the one object and have the other one set up for removal.

**Comment by maddanio** [comment]
Is there then a stable way to completely replace an object?
